subsys: mgmt: Fix broken OTA firmware update

MariuszSkamra opened the pull request:
[DNM] ext: mcumgr: Refactor req & rsp buffer freeing
    
    This refactors the buffer free mechaninsm, so that application
    do not have to free the rsp buffer from tx context.
    
    Relates to https://github.com/apache/mynewt-mcumgr/pull/8
    
    Fixes 8636.
    Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

carlescufi commented:
CC @JoeHut

jhedberg commented on subsys/mgmt/smp_bt.c:
```diff
@@ -51,14 +48,12 @@ static ssize_t smp_bt_chr_write(struct bt_conn *conn,
 
-	ud = net_buf_user_data(nb);
-	ud->conn = bt_conn_ref(conn);
+	bt_smp_conn = conn;
```
This is broken. You have to use bt_conn_ref() whenever making persistent references to a bt_conn (preferably when making any reference)

jhedberg replied:
Also, you're missing the place where you'd release the reference. If you don't get the reference counting right you'll be stuck with a seemingly valid (non-NULL) pointer but you'll never know when it's actually pointing to a valid object. The only way to ensure this is to grab a reference yourself.

nvlsianpu replied:
I wonder whether it is  good idea to take the reference only for offset == 0? AFIK It depends form client side implementation of the long write operation.

jhedberg commented on subsys/mgmt/smp_bt.c:
```diff
@@ -115,30 +96,13 @@ static u16_t smp_bt_get_mtu(const struct net_buf *nb)
-		return 0;
-	}
-
+	conn = bt_smp_conn;
```
No need for a separate local `conn` variable here

jhedberg commented on subsys/mgmt/smp_bt.c:
```diff
@@ -147,15 +111,10 @@ static int smp_bt_tx_pkt(struct zephyr_smp_transport *zst, struct net_buf *nb)
-		rc = smp_bt_tx_rsp(conn, nb->data, nb->len);
-		bt_conn_unref(conn);
-	}
+	conn = bt_smp_conn;
```
No need for a local `conn` variable here. Just use bt_smp_conn directly.

jhedberg reviewed (changes requested):
I also commented in your own tree, since that's what I got an email notification for first. I only later discovered that you had made a pull request as well.

MariuszSkamra committed:
subsys: mgmt: Fix broken OTA firmware update

This fixes freeing net_buf without bt_conn_unref call.
As the result, the OTA was broken.

Fixes 8636
Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra commented:
@jhedberg @nvlsianpu 
I think that conn reference should be per buffer. Managing reference on connected/disconnected is not the right approach in my opinion, because I do not know how it would work in case of multiple connections. 
I've pushed the final version of this patch. I've tested basic image listing, update, listing again, works.
If the user_data is transport specific, the transport has to have an option to free the user_data IMO. 

If you have another vision of this, I will not be mad if you push the patch.

jhedberg commented on subsys/mgmt/smp_bt.c:
```diff
@@ -127,16 +127,28 @@ static u16_t smp_bt_get_mtu(const struct net_buf *nb)
+	struct smp_bt_user_data *ud;
+
+	ud = net_buf_user_data(dst);
+	memcpy(ud, net_buf_user_data(src), sizeof(src->user_data));
```
Since you know that the full user data is of type smp_bt_user_data, I'd avoid this memcpy and just directly do:

```
if (src_ud->conn) {
        dst_ud->conn = bt_conn_ref(src_ud->conn);
}
```

I think that would be the clean way to do this.

nvlsianpu commented:
@MariuszSkamra Looks good to me. I've tested this as well and it works perfectly.
Please fix <b>Shippable.</b>

jhedberg reviewed (changes requested):
This looks good to me, except for the user data copying part that I've commented on. It'd be important to get a code owner to do the final review however.

MariuszSkamra commented:
@ccollins476ad Could you please take a look?

carlescufi commented:
>This looks good to me, except for the user data copying part that I've commented on

@nvlsianpu can you take a look at the part @jhedberg mentions?

nvlsianpu reviewed (approved):
As @jhedberg wrote - it is correct now.
